File: 2802-Find The K-th Lucky Number.py
class Solution:
    def kthLuckyNumber(self, k: int) -> str:
        
        def S(n):
            return 2*(pow(2, n) - 1)
            
        left, right = 0, 30
        while left + 1 < right:
            mid = left + (right-left)//2
            if S(mid) > k:
                right = mid
            else:
                left = mid
                
        largest = left
        remain = k - S(largest)

        if remain == 0:
            return "7"*largest
        else:
            return bin(remain-1)[2:].zfill(largest+1).replace('0', '4').replace('1', '7')

File: C++/2802-Find The K-th Lucky Number.cpp
class Solution {
public:
    string kthLuckyNumber(int k) {
        int n = 1;

        while (k > (1<<n)) {
            k -= (1<<n);
            n++;
        }

        k--;

        deque<char> ls;

        while (n > 0) {
            if (k%2 == 1) ls.push_front('7');
            else ls.push_front('4');
            k = k/2;
            n--;
        }

        return string(ls.begin(), ls.end());
    }
};